pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/readers/ImmutableDictionaryReader.java

```diff
@@ -15,137 +15,183 @@
+    _valueReader = new FixedByteValueReaderWriter(dataBuffer);
+    _length = length;
+    _numBytesPerValue = numBytesPerValue;
+    _paddingByte = paddingByte;
   }
 
-  protected int intIndexOf(int actualValue) {
-    return fileSearcher.binarySearch(0, actualValue);
+  @Override
+  public int length() {
+    return _length;
   }
 
-  protected int floatIndexOf(float actualValue) {
-    return fileSearcher.binarySearch(0, actualValue);
+  @Override
+  public void close() {
+    _valueReader.close();
   }
 
-  protected int longIndexOf(long actualValue) {
-    return fileSearcher.binarySearch(0, actualValue);
+  protected int binarySearch(int value) {
+    int low = 0;
+    int high = _length - 1;
+    while (low <= high) {
+      int mid = (low + high) >>> 1;
+      int midValue = _valueReader.getInt(mid);
+      if (midValue < value) {
+        low = mid + 1;
+      } else if (midValue > value) {
+        high = mid - 1;
+      } else {
+        return mid;
+      }
+    }
+    return -(low + 1);
   }
 
-  protected int doubleIndexOf(double actualValue) {
-    return fileSearcher.binarySearch(0, actualValue);
+  protected int binarySearch(long value) {
+    int low = 0;
+    int high = _length - 1;
+    while (low <= high) {
+      int mid = (low + high) >>> 1;
+      long midValue = _valueReader.getLong(mid);
+      if (midValue < value) {
+        low = mid + 1;
+      } else if (midValue > value) {
+        high = mid - 1;
+      } else {
+        return mid;
+      }
+    }
+    return -(low + 1);
   }
 
-  protected int stringIndexOf(String actualValue) {
-    return fileSearcher.binarySearch(0, actualValue);
+  protected int binarySearch(float value) {
+    int low = 0;
+    int high = _length - 1;
+    while (low <= high) {
+      int mid = (low + high) >>> 1;
+      float midValue = _valueReader.getFloat(mid);
+      if (midValue < value) {
+        low = mid + 1;
+      } else if (midValue > value) {
+        high = mid - 1;
+      } else {
+        return mid;
+      }
+    }
+    return -(low + 1);
   }
 
-  @Override
-  public abstract int indexOf(Object rawValue);
-
-  @Override
-  public abstract Object get(int dictionaryId);
-
-  @Override
-  public abstract long getLongValue(int dictionaryId);
-
-  @Override
-  public abstract double getDoubleValue(int dictionaryId);
+  protected int binarySearch(double value) {
+    int low = 0;
+    int high = _length - 1;
+    while (low <= high) {
+      int mid = (low + high) >>> 1;
+      double midValue = _valueReader.getDouble(mid);
+      if (midValue < value) {
+        low = mid + 1;
+      } else if (midValue > value) {
+        high = mid - 1;
+      } else {
+        return mid;
+      }
+    }
+    return -(low + 1);
+  }
 
-  @Override
-  public int length() {
-    return rows;
+  protected int binarySearch(String value) {
+    byte[] buffer = getBuffer();
+    int low = 0;
+    int high = _length - 1;
+    if (_paddingByte == 0) {
```

Comment:
It seems that we use `0` for checking whether to use unpadded or padded. Can we use `public static final`?

Reply:
The final goal is to use only 0 for padding, and remove all checks. Will keep it like this for now because cannot find a good static name for this 0 :P